src/runtime_src/core/common/memaccess.cpp

```diff
@@ -0,0 +1,216 @@
+
+  auto validated_size = size;
+  // If no size is specified for the read. Read all available memory
+  if ((size == 0) && (action == operation_type::read))
+    validated_size = available_size;
+
+  uint64_t current_addr = validated_start_addr;
+  size_t remaining_bytes_to_see = size;
+  size_t bytes_seen = 0;
+
+  // continue to operate as long as there are bytes left to see or we run out of banks
+  for (auto it = start_bank; (it != vec_banks.end()) && (remaining_bytes_to_see > 0); ++it) {
+    // Validate the amount of memory the current memory bank has
+    uint64_t available_bank_size = 0;
+    if (it != start_bank) {
+      current_addr = it->m_base_address;
+      available_bank_size = it->m_size;
+    }
+    else 
+      available_bank_size = it->m_size - (current_addr - it->m_base_address);
+
+    // If the available bank size is less than the source buffer see what bytes we are able to and move to the next bank
+    uint64_t bytes_to_edit = std::min(available_bank_size, validated_size);
+
+    // Update the buffer index based on how far we have read
+    void* current_buffer_location = static_cast<char *>(buf.get()) + bytes_seen;
+    switch (action) {
+      case operation_type::read:
+        device->unmgd_pread(current_buffer_location, bytes_to_edit, validated_start_addr);
+        break;
+      case operation_type::write:
+        device->unmgd_pwrite(current_buffer_location, bytes_to_edit, validated_start_addr);
+        break;
+      default:
+        throw std::runtime_error("perform_memory_action: Invalid action");
+    }
+    remaining_bytes_to_see -= bytes_to_edit;
+    bytes_seen += bytes_to_edit;
+  }
+
+  if (remaining_bytes_to_see > 0)
+      throw std::runtime_error(boost::str(boost::format("Warning: Saw %llu bytes. Requested %llu bytes") % bytes_seen % size));
+}
+
+} // Empty namespace
+
+namespace xrt_core {
+
+std::vector<char>
+device_mem_read(device* device, const uint64_t start_addr, const uint64_t size)
+{
+  // Allocate a buffer to hold the read data
+  auto buf = xrt_core::aligned_alloc(xrt_core::getpagesize(), size);
+  if (!buf)
+    throw std::runtime_error("read_banks: Failed to allocate aligned buffer");
+  std::memset(buf.get(), 0, size);
+
+  // Read from the device
+  perform_memory_action(device, buf, start_addr, size, operation_type::read);
+
+  // Format the read data into the return object
+  std::vector<char> data(size);
+  std::memcpy(data.data(), buf.get(), size);
+  return data;
+}
+
+void
+device_mem_write(device* device, const uint64_t start_addr, const std::vector<char>& src) 
+{
+  // Prepare the data to write
+  auto buf = xrt_core::aligned_alloc(xrt_core::getpagesize(), src.size());
+  if (!buf)
+    throw std::runtime_error("write_banks: Failed to allocate aligned buffer");
+  std::memcpy(buf.get(), src.data(), src.size());
+
+  // Write to the device
+  perform_memory_action(device, buf, start_addr, src.size(), operation_type::write);
```

Comment:
Hmm....does the buf value have a size.   If so, then src.size() is not needed and this method should be refactored accordingly.

Reply:
buf is a unique pointer so no size is held in there.